vignettes/higher_order.Rmd

```diff
@@ -0,0 +1,205 @@
+
+```{r}
+vs_with <- vs %>%
+  call_derivation(
+    derivation = derive_var_extreme_flag,
+    variable_params = list(
+      params(new_var = AHIFL, mode = "last"),
+      params(new_var = ALOFL, mode = "first")
+    ),
+    by_vars = vars(USUBJID, VSTESTCD),
+    order = vars(VSORRES, VSSEQ)
+  )
+```
+
+```{r, eval=TRUE, echo=FALSE}
+vs_with %>%
+  arrange(USUBJID, VSTESTCD, VSDY, VSSEQ) %>%
+  dataset_vignette(
+    display_vars = vars(USUBJID, VSTESTCD, VSORRES, ALOFL, AHIFL),
+    filter = VSTESTCD %in% c("TEMP", "WEIGHT")
+  )
+```
+
+The benefits of this higher order function would be further highlighted with
+examples where more than two variable derivations had similar needs.
+Developing your ADaM code this way would be more efficient and readable,
+maintenance would be eased in case of specification changes, and downstream
+quality checking would require less effort.
+
+# Restrict Derivation
+
+The idea behind this function is that sometimes you want to apply a
+derivation only for certain records from the input dataset. Introducing
+`restrict_derivation()` therefore gives the users the ability to achieve
+this across any function, without each function needing to have such an
+argument to allow for this.
+
+An example would be if you wanted to flag the first occurring AE with the
+highest severity for each patient, BUT you only wanted to do this for
+records occurring on or after study day 1.
+
+Here is how you could achieve this using `restrict_derivation()`,
+where the function arguments are passed using `params()` and the restriction
+criteria is given using `filter`:
+
+```{r}
+ae <- ae %>%
+  mutate(TEMP_AESEVN = as.integer(factor(AESEV, levels = c("SEVERE", "MODERATE", "MILD")))) %>%
+  restrict_derivation(
+    derivation = derive_var_extreme_flag,
+    args = params(
+      new_var = AHSEVFL,
+      by_vars = vars(USUBJID),
+      order = vars(TEMP_AESEVN, AESTDY, AESEQ),
+      mode = "first"
+    ),
+    filter = AESTDY >= 1
+  )
+```
+
+```{r, eval=TRUE, echo=FALSE}
+ae %>%
+  arrange(USUBJID, AESTDY, AESEQ, desc(TEMP_AESEVN)) %>%
+  dataset_vignette(
+    display_vars = vars(USUBJID, AEDECOD, AESTDY, AESEQ, AESEV, AHSEVFL)
+  )
+```
+
+# Slice Derivation
+
+This function in a way combines the features of the above two. It allows a
+derivation to be applied with different arguments for different slices (subsets)
+of records from the input dataset. You could do this with separate `restrict_derivation()`
+calls for each different set of records, but `slice_derivation()` allows
+to achieve this in one call.
+
+An example would be if you wanted to achieve the same derivation as above
+for records occurring on or after study day 1, BUT for pre-treatment AEs
+you wanted to flag only the last occurring AE.
+
+Here is how you could achieve this using `slice_derivation()`,
+where the function arguments are passed using `params()` and via the different
+slices controlled by `filter`:
```

Comment:
What I'm missing here is an explanation of `TRUE` being passed to `filter`. ALso an additional examle with three different slices might be good to showcase that the order of the slices matters (if I recall correctly, please correct @bundfussr).

Reply:
@thomas-neitmann , you are correct. The order matters if the conditions are not mutually exclusive (see "Details" section of `slice_derivation()`).